ignite/handlers/param_scheduler.py

```diff
@@ -1015,7 +1019,7 @@ def print_lr():
             if init_lr != param_group_warmup_end_value:
                 milestones_values.append((warmup_duration, init_lr))
 
-            lr_scheduler = LRScheduler(lr_scheduler, save_history=save_history)
+            lr_scheduler = LRScheduler(lr_scheduler, save_history=save_history, use_legacy=True)
```

Comment:
Can we check if we could use `use_legacy=False` ? Or what is blocking here ?

Reply:
if `use_legacy=False`,  the result look like the example below
- example1
```python
from torch.optim.lr_scheduler import ExponentialLR

tensor = torch.zeros([1], requires_grad=True)
optimizer = torch.optim.SGD([tensor], lr=0.01)
torch_lr_scheduler = StepLR(optimizer, step_size=3, gamma=0.1)

trainer = Engine(dummy_update)
scheduler = create_lr_scheduler_with_warmup(torch_lr_scheduler,
                                            warmup_start_value=0.0,
                                            warmup_end_value=0.1,
                                            warmup_duration=3)

trainer.add_event_handler(Events.ITERATION_STARTED, scheduler)

@trainer.on(Events.ITERATION_COMPLETED)
def print_lr():
    print(optimizer.param_groups[0]["lr"])

_ = trainer.run([0] * 8, max_epochs=1)
```
- output1
```
0.0
0.05
0.1
0.01
0.01
0.01
0.01
0.001
```
- example2
```python
from torch.optim.lr_scheduler import ExponentialLR

tensor = torch.zeros([1], requires_grad=True)
optimizer = torch.optim.SGD([tensor], lr=0.1)
torch_lr_scheduler = StepLR(optimizer, step_size=3, gamma=0.1)

trainer = Engine(dummy_update)
scheduler = create_lr_scheduler_with_warmup(torch_lr_scheduler,
                                            warmup_start_value=0.0,
                                            warmup_end_value=0.1,
                                            warmup_duration=3)

trainer.add_event_handler(Events.ITERATION_STARTED, scheduler)

@trainer.on(Events.ITERATION_COMPLETED)
def print_lr():
    print(optimizer.param_groups[0]["lr"])

_ = trainer.run([0] * 8, max_epochs=1)
```
- output2
```
0.0
0.05
0.1
0.1
0.1
0.1
0.010000000000000002
0.010000000000000002
```
The initial value of stepLR (= 0.1 or 0.01) is counted one extra time than step_size (=3).

Reply:
Thanks for the details @yuta0821 , can you please debug this a bit more and explicitly say which scheduler is responsible for adding LR value (0.1 or 0.01). I'm not quite sure to understand why exactly this happens. Thanks !

Reply:
@yuta0821 sorry for delay and thanks for the explanation! I haven't checked it in details but will do as soon as it could be possible from my side (~4-5 days).
@sdesrozis can you help with that if you have some bandwidth ?

Reply:
@vfdev-5 
Thank you for your reply, there is no problem.

Reply:
@vfdev-5 Ok I will check

Reply:
@yuta0821 The point is `use_legacy=True` mustn’t be used internally because we will depreciate in the futur. All the things should work correctly using the new behaviour.

Anyway, have you checked the ConcatScheduler ? It seems the concatenation should not work as excepted without `use_legacy=True`, right ?

Reply:
> @yuta0821 The point is use_legacy=True mustn’t be used internally because we will depreciate in the futur. All the things should work correctly using the new behaviour.

Ok, I understand.

> Anyway, have you checked the ConcatScheduler ? 

Yes, I have checked the `ConcatScheduler`.  
> It seems the concatenation should not work as excepted without use_legacy=True, right ?

The 'excepted' is 'expected' ? If `use_legacy` is True, the initial `lr` value of optimizer is repeated extra once.  
I think the concatenation between `LRScheduler` and `PiecewiseLinear` in `ConcatScheduler` causes unexpected behavior.

Reply:
It performs the same operation as `use_legacy`, but wouldn't it be preferable to add the argument `skip_initial_value` as a variable to be used for the internal function ?

Reply:
Yeah, expected rather than excepted 😅

As far I understand, using `use_legacy=False`, the first `lr` comes from the optimizer. Whatever schedulers used, the schedulers concatenation will produce a repetition at each joint. 

Having an internal option as you suggested sounds good to me. I mean rename `use_legacy` to `skip_initial_value` is fine. Although, we have to keep `use_legacy` for the users.

What do you think ?

Reply:
@sdesrozis 
I am sorry for having consulted with you.
It seems that this problem can be solved by setting the internal variable `keep_first_value=True` only when `create_lr_scheduler_with_warmup` is called, as shown below, to store the initial value in `LRScheduler`.
```python
class LRScheduler(ParamScheduler):
    def __init__(self, lr_scheduler: _LRScheduler, save_history: bool = False, use_legacy: bool = False, keep_first_lr: bool = False):
        if keep_first_lr:
            self.lr_scheduler._get_lr_called_within_step = True  # type: ignore[attr-defined]
            self.first_lr = self.lr_scheduler.get_lr()
            self.lr_scheduler._get_lr_called_within_step = False  # type: ignore[attr-defined]

    def get_param(self) -> Union[float, List[float]]:
        """Method to get current optimizer's parameter value"""
        # Emulate context manager for pytorch>=1.4
        if hasattr(self, "first_lr"):
            lr_list = self.first_lr
            del self.first_lr
        else:
       
def create_lr_scheduler_with_warmup( ):
    if isinstance(lr_scheduler, _LRScheduler):
        init_lr = param_group["lr"]
        lr_scheduler = LRScheduler(lr_scheduler, save_history=save_history, keep_first_lr=True)
    else:
        init_lr = lr_scheduler.get_param()
```
I am running the existing test now. I will commit once all tests pass!

Reply:
It seems that the tests are ko...

Reply:
Sorry, I added `type: ignore[attr-defined]` !